DataFormats/Portable/interface/PortableHostCollectionReadRules.h

```diff
@@ -0,0 +1,83 @@
+#ifndef DataFormats_Portable_interface_PortableHostCollectionReadRules_h
+#define DataFormats_Portable_interface_PortableHostCollectionReadRules_h
+
+#include <boost/core/demangle.hpp>
+
+// required by <RtypesImp.h>
+#define G__DICTIONARY
+
+#include <Rtypes.h>
+#include <RtypesImp.h>
+#include <TGenericClassInfo.h>
+#include <TVirtualObject.h>
+
+#include "DataFormats/Portable/interface/PortableHostCollection.h"
+#include "FWCore/Utilities/interface/concatenate.h"
+#include "FWCore/Utilities/interface/stringize.h"
+
+// read function for PortableHostCollection, called for every event
+template <typename T>
+static void readPortableHostCollection(char *target, TVirtualObject *from_buffer) {
+  // extract the actual types
+  using Collection = T;
+  using Layout = typename Collection::Layout;
+
+  // valid only for PortableHostCollection<T>
+  static_assert(std::is_same_v<Collection, PortableHostCollection<Layout>>);
+
+  // proxy for the object being read from file
+  struct OnFile {
+    Layout &layout_;
+  };
+
+  // address in memory of the buffer containing the object being read from file
+  char *address = static_cast<char *>(from_buffer->GetObject());
+  // offset of the "layout_" data member
+  static ptrdiff_t layout_offset = from_buffer->GetClass()->GetDataMemberOffset("layout_");
```

Comment:
@pcanal General question: can this offset change between files that are being read by one process, if one of the files has a different version of the class begin read where the offset to the data member is different compared to the class versions in the other files?

We haven't explored the schema evolution for the SoA or PortableHostCollection at all (and our current stance is that the schema evolution should not be assumed to work), so _in practice_ the `layout_offset` being `static` does not make the situation worse _for now_.

Reply:
From my understanding of the approach used internally by ROOT, if we decide to support schema evolution we should have a different `readPortableHostCollection` for each version, with different rules referring to them in `set_PortableHostCollection_read_rules`.

Maybe I should rename `readPortableHostCollection` to `readPortableHostCollection_v1`.

Reply:
> From my understanding of the approach used internally by ROOT, if we decide to support schema evolution we should have a different `readPortableHostCollection` for each version, with different rules referring to them in `set_PortableHostCollection_read_rules`.

Thanks @fwyzard. I suppose having separate function templates would deal with the evolution of the `PortableHostCollection<TLayout>` itself. The other part is then the evolution of the SoA definition, i.e. `TLayout` (that I'd expect to be much more common case).

Written all that, I think we can, for now, continue with the statement "schema evolution is not supported". But at some point the schema evolution details should be explored.

Reply:
Yes, you would have one such function per old schema.   You probably would not need to have multiple in memory representation of the old schema (In most case you don't).

Reply:
< static ptrdiff_t layout_offset = from_buffer->GetClass()->GetDataMemberOffset("layout_");

This is the location in memory of the target in the "current" schema is completely independent of the content of the ROOT files .. and completely tied to the content of the shared library.   Depending of the content of the in-memory class, this may or may not change with the instances.

Reply:
> static ptrdiff_t layout_offset = from_buffer->GetClass()->GetDataMemberOffset("layout_");

> can this offset change between files that are being read by one process, if one of the files has a different version of the class begin read where the offset to the data member is different compared to the class versions in the other files?

Technically yes.  This depends on the version of the data schema layout being read.